HARK/ConsumptionSaving/ConsRiskyAssetModel.py

```diff
@@ -0,0 +1,1944 @@
+            
+            # Loop over each period of the cycle, getting controls separately depending on "age"
+            for t in range(self.T_cycle):
+            
+                # Find agents in this period-stage
+                these = t == self.t_cycle
+                
+                if np.sum(these) > 0:
+                    tau = self.tau[t]
+                    
+                    mNrmTildeNow[these], nNrmTildeNow[these] = rebalanceAssets(self.controls['DNrmNow'][these],
+                                                                               self.state_now['mNrmNow'][these],
+                                                                               self.state_now['nNrmNow'][these],
+                                                                               tau)
+        
+        self.state_now['mNrmTildeNow'] = mNrmTildeNow
+        self.state_now['nNrmTildeNow'] = nNrmTildeNow
+    
+    def getControlsSha(self):
+        """
+        """
+        
+        ShareNow = np.zeros(self.AgentCount) + np.nan
+        
+        # Loop over each period of the cycle, getting controls separately depending on "age"
+        for t in range(self.T_cycle):
+            
+            # Find agents in this period-stage
+            these = t == self.t_cycle
+                           
+            # Get controls for agents who *can* adjust.
+            those = np.logical_and(these, self.shocks['AdjustNow'])
+            ShareNow[those] = self.solution[t].stageSols['Sha'].ShareFuncAdj(
+                self.state_now['mNrmTildeNow'][those],
+                self.state_now['nNrmTildeNow'][those]
+            )
+                
+            # Get Controls for agents who *can't* adjust.
+            those = np.logical_and(these, np.logical_not(self.shocks['AdjustNow']))
+            ShareNow[those] = self.solution[t].stageSols['Sha'].ShareFuncFxd(
+                self.state_now['mNrmTildeNow'][those],
+                self.state_now['nNrmTildeNow'][those],
+                self.state_now['ShareNow'][those]
+            )
+
+        # Store controls as attributes of self
+        self.controls['ShareNow'] = ShareNow     
+        # Share is also a state
+        # TODO: Ask Seb how this is handled
+        self.state_now['ShareNow'] = ShareNow
+        
+    def getStatesCns(self):
+        # No new states need to be computed in the consumption stage
+        pass
+        
+    def getControlsCns(self):
+        """
+        """
+        
+        cNrmNow = np.zeros(self.AgentCount) + np.nan
+        
+        # Loop over each period of the cycle, getting controls separately depending on "age"
+        for t in range(self.T_cycle):
+            
+            # Find agents in this period-stage
+            these = t == self.t_cycle
+                           
+            # Get consumption
+            cNrmNow[these] = self.solution[t].stageSols['Cns'].cFunc(
```

Comment:
Notice the structural similarity between this and the other getControls methods.
The `for` loop is doing the same work in each one.
If the shared logic could be refactored out such that each individual update was a special case of a more general method, that would be a step towards developing a general solution _algorithm_ (as opposed to writing code that reflects a particular set of mathematical statements).

Reply:
This part deals with simulation. Notice that `cFunc` is already defined.

I believe this might be generalizable to some `applyPolicyFunc` method.

This could work by keeping policy functions in a dictionary, say, self.solution[t].PolFuncs = {'c': ...}

One would need to keep track of their arguments. The nice thing is that their arguments should always be states, so those are also in dictionaries already, I believe.